### ISSUE TITLE
Docker setup fails with Git Error

### ISSUE THREAD

**Issue by belibug** [Linbreux/wikmd#42]
Running wikmd as docker container with following error.

```
wikmd    | [2022-05-28 19:35:41,212] INFO in git_manager: Initializing existing repo >>> wiki ...
wikmd    | [2022-05-28 19:35:41,212] INFO in git_manager: Creating 'main' branch ...
wikmd    | [2022-05-28 19:35:41,216] ERROR in git_manager: Existing repo initialization failed >>> Cmd('git') failed due to: exit code(128)
wikmd    |   cmdline: git branch -M main
wikmd    |   stderr: 'fatal: unsafe repository ('/app/wikmd/wiki' is owned by someone else)
wikmd    | To add an exception for this directory, call:
wikmd    |
wikmd    |      git config --global --add safe.directory /app/wikmd/wiki'
```
Troubleshooting done,
1. Configured PID GID to current user
2. Ensured path/to/wiki has correct permission

**Comment by Linbreux** [comment]
Hi @belibug , 
the docker image has known issues. I'm working on it to resolve it. For now I would suggest running it native on your machine if possible to get the best experience.

**Comment by Waldorf3** [comment]
> Hi @belibug , the docker image has known issues. I'm working on it to resolve it. For now I would suggest running it native on your machine if possible to get the best experience.

I hope you'll be able to upload a working docker image soon.  Wikmd seems to be exactly what I've been looking for - simple clean interface, no nonsense, not trying to be a lot of other things.

**Comment by Linbreux** [comment]
Hi @Waldorf3, I tried looking for the problem, but I can't seem to find how to get the user permission to work. 

@artivis Sorry for the inconvenience, your Dockerfiles worked like a charm. But do you know how to fix the permission issue here, I can't seem to find the problem?

**Comment by artivis** [comment]
Hey, no problem. A quick search seem to point to a [recent change in git](https://github.blog/2022-04-12-git-security-vulnerability-announced/). It has affected a lot of other projects. I'm not at my computer so I can't debug that right now but the permissions are set [here](https://github.com/Linbreux/wikmd/blob/main/docker/root/etc/cont-init.d/30-config#L16) for new repo created from the example. I'll give it a look on Monday :+1:.

**Comment by Linbreux** [comment]
Alright, I'll look into it. Thanks for the quick response, appreciate it!

**Comment by Waldorf3** [comment]
> Hi @Waldorf3, I tried looking for the problem, but I can't seem to find how to get the user permission to work. It seems like the docker/root/etc/cont-init.d/30-config file does not work.
> 
> I fixed it for now using the following method:
> 
> ```
> docker exec -it <container-id> bash
> chown -R abc:abc wiki/ 
> chown -R abc:abc wikmd
> ```
> 
> let me know if that worked ;)
> 
> @artivis Sorry for the inconvenience, your Dockerfiles worked like a charm. But do you know how to fix the permission issue here, I can't seem to find the problem?

/wiki already is  www-data.www-data (my preferred uid-gid)
There is no /wikmd. There is a /app/wikmd, and if i chown that as well it kinda works, but data is now stored under /app/wikmd/wiki - not /wiki .  I guess the quick workaround would be to mount /app/wikmd/wiki in the docker-compose instead?

**Comment by Linbreux** [comment]
My excuses, I meant 'app/wikmd'. How does your docker-compose look like? I also tried to chown the mount folder on my host machine with the hosts user (even if it already was the same) and that also made it work.

If you don't specify a volume the files won't be stored in 'wiki'.

**Comment by Waldorf3** [comment]
My docker-compose.yml is pretty much straight from the github:

```
version: "2.1"
services:
  wikmd:
    image: linbreux/wikmd:latest
    container_name: wikmd
    environment:
      - PUID=33
      - PGID=33
      - TZ=Europe/Brussels
#      - HOMEPAGE=homepage.md #optional
#      - HOMEPAGE_TITLE=homepage.md #optional
#      - WIKMD_LOGGING=1 #optional
    volumes:
      - /homes/me/wiki:/app/wikmd/wiki
    ports:
      - 8086:5000
    restart: unless-stopped
```
btw, would it not make sense to have a config folder on the host for whatever users might customize (template etc)

**Comment by Linbreux** [comment]
Could you try: 
```
---
version: "2.1"
services:
  wikmd:
    image: linbreux/wikmd:latest
    container_name: wikmd
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Paris
      - HOMEPAGE=homepage.md #optional
      - HOMEPAGE_TITLE=homepage.md #optional
      - WIKMD_LOGGING=1 #optional
      - WIKI_DIRECTORY=/wiki
    volumes:
      - /home/me/wiki:/wiki
    ports:
      - 8086:5000
    restart: unless-stopped

```
Make sure to specify the **WIKI_DIRECTORY**. Also try to chown `/home/me/wiki` again on the host.

**Comment by Waldorf3** [comment]
Yes that worked, thanks!

The documentation files are no longer present though, also not when hitting list as before, is that on purpose?

I get your point about config through env variables, makes good sense, sorry I didn't think about that.

**Comment by Linbreux** [comment]
Hi good to hear that, also make sure to create a img file in the wiki folder when you want to upload images through the webinterface.

The demo files are indeed not present, thats a problem I'll fix in the future. But for know everything should work.

**Comment by Waldorf3** [comment]
I copied the content from /app/wikmd/wiki to /wiki, that seems to take are of it?

**Comment by artivis** [comment]
I couldn't reproduce the error described here (did find a couple other tho).
@belibug any chance you could provide a minimal reproducible example?

**Comment by belibug** [comment]
@artivis this is how you can reproduce the error.
I used below docker compose yaml file to build docker image. 
```
version: "2.1"
services:
  wikmd:
    image: linbreux/wikmd:latest
    container_name: wikmd
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Europe/Paris
      - HOMEPAGE=homepage.md #optional
      - HOMEPAGE_TITLE=homepage.md #optional
      - WIKMD_LOGGING=1 #optional
    volumes:
      - ./wiki:/wiki
    ports:
      - 5000:5000
    restart: unless-stopped
```
As you can see I mapped local immediate path 'wiki' to 'wiki' and I get below error.

```
wikmd    | INFO in git_manager: Creating a new local repo >>> wiki ...
wikmd    | ERROR in git_manager: New local repo initialization failed >>> Cmd('git') failed due to: exit code(1)
wikmd    |  cmdline: git init
wikmd    |  stderr: '/app/wikmd/wiki/.git: Permission denied'
```

I have tried mapping volume as `/wiki:/app/wikmd/wiki' with no success. Like others I tried chown the folder and set to full open permission `777` which throws new error as below.

```
wikmd    | INFO in git_manager: Creating a new local repo >>> wiki ...
wikmd    |  INFO in git_manager: Creating 'main' branch ...
wikmd    |  ERROR in git_manager: New local repo initialization failed >>> Cmd('git') failed due to: exit code(128)
wikmd    |  cmdline: git branch -M main
wikmd    |  stderr: 'fatal: unsafe repository ('/app/wikmd/wiki' is owned by someone else)
wikmd    | To add an exception for this directory, call:
wikmd    |
wikmd    |   git config --global --add safe.directory /app/wikmd/wiki'
wikmd    |  * Serving Flask app "wiki" (lazy loading)
```

Thank you.

**Comment by artivis** [comment]
Hi @belibug, 
Thanks for the detailed report. Could you try the fix I proposed in #45 :

- build the image,

```bash
git clone https://github.com/artivis/wikmd.git -b fix/docker
cd wikmd
docker build -t linbreux/wikmd:fix -f docker/Dockerfile .
```

- launch the container,

```
docker run -d \
  --name wikmd \
  -e TZ=Europe/Paris \
  -e PUID=1000 \
  -e PGID=1000 \
  -e HOMEPAGE=homepage.md `#optional` \
  -e HOMEPAGE_TITLE=homepage.md `#optional` \
  -e WIKMD_LOGGING=1 `#optional` \
  -p 5000:5000 \
  -v ./wiki:/wiki \
  --restart unless-stopped \
  linbreux/wikmd:fix
```

Make sure that the permissions are correctly set, e.g. with,

```bash
$ id myuser
uid=1001(myuser) gid=1002(myuser) groups=1001(myuser)
```

then you should set `PUID=1001` and `PGID=1002`.

**Comment by belibug** [comment]
@artivis  I tried building from docker file, It worked for most part but failed to rename branch which is same as issue #44 which I see you opened. 

```
wikmd    | [2022-06-09 03:37:18,438] INFO in git_manager: Initializing existing repo >>> /wiki ...
wikmd    | INFO in git_manager: Creating 'main' branch ...
wikmd    | ERROR in git_manager: Existing repo initialization failed >>> Cmd('git') failed due to: exit code(128)
wikmd    | cmdline: git branch -M main
wikmd    | stderr: 'error: refname refs/heads/master not found
wikmd    | fatal: Branch rename failed'
```

**Comment by Linbreux** [comment]
hi @belibug 

This problem should be solved in 4c659cad28f8b96c02ffcc7c615b85083aaa1ab8.
